src/ucp/core/ucp_ep.c

```diff
@@ -836,8 +859,23 @@ ucs_status_ptr_t ucp_ep_close_nb(ucp_ep_h ep, unsigned mode)
                                     UCT_FLUSH_FLAG_LOCAL : UCT_FLUSH_FLAG_CANCEL,
                                     NULL, 0, NULL,
                                     ucp_ep_close_flushed_callback, "close");
+
     if (!UCS_PTR_IS_PTR(request)) {
-        ucp_ep_disconnected(ep, mode == UCP_EP_CLOSE_MODE_FORCE);
+        uct_cm_ep = ucp_ep_get_cm_uct_ep(ep);
+        if (uct_cm_ep != NULL) {
+            if (ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
+                ucp_ep_cm_disconnect_cm_lane(ep);
+                close_req = ucp_ep_cm_close_request_get(ep);
+                ucp_ep_set_close_request(ep, close_req);
+                request = (close_req != NULL) ? close_req + 1 :
+                          UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+            }
+        }
+    }
+
+    if (!UCS_PTR_IS_PTR(request)) {
+        ucp_ep_disconnected(ep, (mode == UCP_EP_CLOSE_MODE_FORCE) ||
```

Comment:
this should be in 'else' of the 'if' in lines 865-866
if we failed to allocate close request should return ERR_NO_MEMORY, not fallback here and set ep to disconnected

Reply:
but this is `ucp_ep_close_nb`, do we expect user will call it again? otherwise we will have a memory leak
IMO if `ucp_ep_close_nb` fails we have to do all possible to release resources gracefully

Reply:
we should only release resources allocated during ucp_close_nb

Reply:
but what about the rest? do we expect user will call `ucp_ep_close_nb` again, let's say, in force mode?

Reply:
in this case we have no way to really avoid a leak

Reply:
why?
ucp_ep_disconnected with true-value in second arg does it, user will get error status. So, I don't see the reason don't clean up local resources in case of failure

Reply:
what about pending/uncompleted requests?

Reply:
also, it can cause fabric error on the peer if we just destroy the ep

Reply:
ok, will revert for now since common err flow is not related to this PR